File: Unity/SimpleDemoRL/Assets/Scripts/client/api/messages/ObservationMessage.cs
using UnityEngine;

[System.Serializable]
public class EnvState
{
    public Vector3 agentPostion;
    public Vector3 agentRotation;
    public Vector3 velocity;
    public Vector3 angularVelocity;
    public Vector3 redBallPosition;
    public Vector3 blueBallPosition;
    public Vector3 greenBallPosition;
    public Vector3 grayAreaPosition;
    public Vector3 orangeAreaPosition;
    public Vector3 whiteAreaPosition;
    public float agentRedBallAngle;
    public float agentGrayAreaAngle;
    public float agentRedBallDist;
    public float agentGrayAreaDist;
    public float RedBallGreyAreaDist;
    public bool agentOutsidePlane;
    public bool redBallOutsidePlane;
    public bool terminate;
    public float reward;
    public int envNum;


    public static EnvState FromJson(string json)
    {
        return JsonUtility.FromJson<EnvState>(json);
    }

    public string ToJson()
    {
        return JsonUtility.ToJson(this);
    }
}

File: python/messages/observation_message.py
import json

from .message import Message


class ObservationMessage(Message):
    """
    Message specifying an environment state
    """
    def __init__(self, agentPostion, agentRotation, velocity, angularVelocity, 
                 redBallPosition, blueBallPosition, greenBallPosition, 
                 grayAreaPosition, orangeAreaPosition, whiteAreaPosition,
                 agentRedBallAngle, agentGrayAreaAngle, agentRedBallDist,
                 agentGrayAreaDist, RedBallGreyAreaDist, agentOutsidePlane,
                 redBallOutsidePlane,terminate, reward, envNum):
        
        self.agentPostion = agentPostion
        self.agentRotation = agentRotation
        
        self.velocity = velocity
        self.angularVelocity = angularVelocity
        
        self.redBallPosition = redBallPosition
        self.blueBallPosition = blueBallPosition
        self.greenBallPosition = greenBallPosition
        
        self.grayAreaPosition = grayAreaPosition
        self.orangeAreaPosition = orangeAreaPosition
        self.whiteAreaPosition = whiteAreaPosition

        self.agentRedBallAngle = agentRedBallAngle
        self.agentGrayAreaAngle = agentGrayAreaAngle
        self.agentRedBallDist = agentRedBallDist
        self.agentGrayAreaDist = agentGrayAreaDist
        self.RedBallGreyAreaDist = RedBallGreyAreaDist
        self.agentOutsidePlane = agentOutsidePlane
        self.redBallOutsidePlane = redBallOutsidePlane
        
        self.terminate = terminate
        self.reward = reward

        self.envNum = envNum
